### ISSUE TITLE
Drop "cpio" libraries and write something semi-custom, because RPM doesn't use vanilla CPIO

### ISSUE THREAD

**Issue by dralley** [rpm-rs/rpm#108]
See the "Payload" section of the website: https://rpm-software-management.github.io/rpm/manual/format.html

> # Payload
> 
> The Payload is currently a cpio archive, gzipped by default. The cpio archive type used is SVR4 with a CRC checksum.
> 
> As cpio is limited to 4 GB (32 bit unsigned) file sizes RPM since version 4.12 uses a stripped down version of cpio for packages with files > 4 GB. This format uses 07070X as magic bytes and the file header otherwise only contains the index number of the file in the RPM header as 8 byte hex string. The file metadata that is normally found in a cpio file header - including the file name - is completely omitted as it is stored in the RPM header already.

So, we should fork [`cpio-rs`](https://github.com/jcreekmore/cpio-rs) (providing the appropriate credits of course), strip it down to the subset we need, and change the magic bytes constant.

Luckily the CPIO format is pretty simple and the library only a few hundred lines, so it's not a big deal.

Subsequently we need to change the `PAYLOADFORMAT` tag, but upstream RPM still uses `cpio` as the name, so we'll have to wait until they pick something.

**Comment by drahnr** [comment]
A real concern: Do we want to support rpms with cpio-like archives _larger_ than 4GB? It feels like we pull in a lot of pain for supporting an antipattern? Are there use-cases that are _idiomatic_ that require rpms larger than 4GB?

**Comment by dralley** [comment]
@drahnr The example that typically comes up is games, which often include many large assets, or ML models, or their training data.  In practice those are rarely distributed as system packages but it is possible and has been done.

**Comment by drahnr** [comment]
Tbh, I'd prefer we create a separate `rpm-cpio` in the org, rather than moving it into the codebase, and just replace the dependency. Does that sound fair?

**Comment by newpavlov** [comment]
I also think that a separate crate would be a better approach. Maybe you should create a repository for it?

**Comment by dralley** [comment]
I'm a bit lukewarm on having a separate crate, because I can't think of anything apart from an RPM parser which would want to parse RPM payloads.  So it would be a separate crate that we would be the only users of, probably ever.

**Comment by drahnr** [comment]
I am mostly thinking operationally: applying upstream changes would be as easy as a git rebate or merge. I couldn't care less if we stay the only user if it simplifies the maintenence

**Comment by dralley** [comment]
I don't think there will be any maintenance, the library is "finished" and hasn't seen any commits in a year.  CPIO is very simple so there are unlikely to be any bugs.

**Comment by drahnr** [comment]
We haven't reached a conclusion here, my preference is still on forking to `rpm-rs/cpio-rpm` and using that.

**Comment by dralley** [comment]
I still have the opposite preference, tbh.  It's very difficult for me to imagine the supposed maintenance benefit repaying itself against having a new crate which nobody but this particular library will ever use.

Since the new payload format removes nearly all of the metadata from the archive (because it's duplicated in the RPM header), you can do very little with the payload without also reading the RPM header.

There is also one new development since we last had the discussion though, which is that RPMv6 plans to use only the "new" payload scheme, so it won't be relegated to just packages with files >4gb anymore, it will eventually be all packages.

Mentioned under the "Payload" section here: https://github.com/rpm-software-management/rpm/discussions/2374

**Comment by dralley** [comment]
Note to self: make sure we handle the edge cases discussed here: https://blog.colindou.ch/posts/lets-make-an-os-cpio-weirdness/

That is: be careful with directories, and be careful with absolute paths. Extracting a cpio shouldn't automatically overwrite /usr/bin/blah.

**Comment by udoprog** [comment]
I'm trying to install an rpm I generated using this crate and I'm getting this on Fedora 41 when I try to install it:
```
[RPM] unpacking of archive failed on file /usr/share/applications/se.tedro.OnTV.desktop;67cf1f3a: cpio: Unknown file type
```
Would this happen to be related?

**Comment by dralley** [comment]
I don't think so, no.  Not unless that particular file is >4gb, but if it was I would think something else would go wrong first?
